[EN-158] document merge logic (#330)
compose-spec/compose-spec: merge.md

## Merge and override

> Compose implementations SHOULD allow users to define a Compose application model through multiple compose files. 

Reviewer: ```suggestion
Compose implementations SHOULD allow users to define a Compose application model through multiple Compose files. 
```

    key3: value3
```

MUST result in a Compose application model equivalent to the YAML tree:

```yaml
services:
  foo:
    key1: value1
    key2: VALUE
    key3: value3
```

### Sequence

A YAML `sequence` is merged by appending values from the overriding Compose file to the previous one.

Merging the following example YAML trees:
```yaml
services:
  foo:
    DNS:
      - 1.1.1.1
```    

```yaml
services:
  foo:
    DNS: 
      - 8.8.8.8
```

MUST result in a Compose application model equivalent to the YAML tree:

```yaml
services:
  foo:
    DNS:
      - 1.1.1.1
      - 8.8.8.8
```

## Exceptions

There are exceptions to those rules:

### Shell commands

Service's [command](#command), [entrypoint](#entrypoint) and [healthcheck](#healthcheck) `test`: 
For usability, the value MUST be overridden by the latest Compose file, and not appended.

Merging the following example YAML trees:
```yaml
services:
  foo:
    command: ["echo", "foo"]
```    

```yaml
services:
  foo:
    command: ["echo", "bar"]
```

MUST result in a Compose application model equivalent to the YAML tree:

```yaml
services:
  foo:
    DNS:
      command: ["echo", "bar"]
```

### Unique resources

> Service's [ports](#ports), [volumes](#volumes), [secrets](#secrets) and [configs](#configs). While those
> are modeled in a Compose file as a sequence, they MUST be unique regarding the target resource in container:

Reviewer: ```suggestion
Applies to service [ports](#ports), [volumes](#volumes), [secrets](#secrets) and [configs](#configs).

While these types are modeled in a Compose file as sequences, they have special uniqueness requirements:
```

| attribute   | unique key               |
|-------------|--------------------------|
| volumes     |  target                  |
| secrets     |  source                  |
| configs     |  source                  |
| ports       |  {ip, target, published, protocol}   |

> While merging Compose files, a Compose implementation MUST respect these unique keys and merge the entries
> in sequence just like this would have been a mapping.

Reviewer: This is a bit hard to interpret. I tried to write an alternative but not claiming it's easy to understand either 🤪 

```suggestion
While merging Compose files, a Compose implementation MUST append new entries that do not violate a uniqueness constraint and merge entries that share a unique key.
```

Author: Yes, I also got issue to explain this. I wonder we should introduce support for a mapping-style syntax on ports, volumes, secrets, config so that this "unicity" concept is more obvious. I will prepare a PR for this purpose

## Merge and override

> Compose Implementations SHOULD support user to define multiple compose file as definition for a Compose

Reviewer 2: ```suggestion
Compose implementations SHOULD support users to define multiple compose files as definition for a Compose
```

> Compose implementations SHOULD support users to define multiple compose files as definition for a Compose

Reviewer 3: ```suggestion
Compose implementations SHOULD allow users to define a Compose application model through multiple Compose files
```

> application model. When doing so, a Compose implementation will merge Compose files alltogether, and MUST

Reviewer 3: ```suggestion
application model. When doing so, a Compose implementation MUST follow the rules declared in this section to merge the Compose files.
```

> follow rules declared in this section.

Reviewer 2: ```suggestion
follow the rules declared in this section.
```

### Mapping

> A YAML `mapping` get merged by adding missing entries and merging the conflicting ones.

Reviewer 2: ```suggestion
A YAML `mapping` gets merged by adding missing entries and merging the conflicting ones.
```

> Merging those two example yaml trees:

Reviewer 2: ```suggestion
Merging those two example YAML trees:
```

> Merging those two example YAML trees:

Reviewer 3: ```suggestion
Merging the following example YAML trees:
```

```yaml
services:
  foo:
    key1: value1
    key2: value2
```    

```yaml
services:
  foo:
    key2: VALUE
    key3: value3
```

> MUST result in a Compose application model equivalent to yaml tree:

Reviewer 2: ```suggestion
MUST result in a Compose application model equivalent to the YAML tree:
```

```yaml
services:
  foo:
    key1: value1
    key2: VALUE
    key3: value3
```

### Sequence

A YAML `sequence` is merged by appending values from the overriding Compose file to the previous one.

> Merging those two example yaml trees:

Reviewer 2: ```suggestion
Merging those two example YAML trees:
```

> Merging those two example YAML trees:

Reviewer 3: ```suggestion
Merging the following example YAML trees:
```

```yaml
services:
  foo:
    DNS:
      - 1.1.1.1
```    

```yaml
services:
  foo:
    DNS: 
      - 8.8.8.8
```

> MUST result in a Compose application model equivalent to yaml tree:

Reviewer 2: ```suggestion
MUST result in a Compose application model equivalent to the YAML tree:
```

```yaml
services:
  foo:
    DNS:
      - 1.1.1.1
      - 8.8.8.8
```

## Exceptions

> There's exceptions to thoes rule:

Reviewer 2: ```suggestion
There are exceptions to those rules:
```

> ### shell commands

Reviewer 4: ```suggestion
### Shell commands
```

Service's [command](#command), [entrypoint](#entrypoint) and [healthcheck](#healthcheck) `test`: 

> For usability, value MUST be overridden by latest Compose file, and not appended.

Reviewer 2: ```suggestion
For usability, the value MUST be overridden by the latest Compose file, and not appended.
```

> Merging those two example yaml trees:

Reviewer 2: ```suggestion
Merging those two example YAML trees:
```

```yaml
services:
  foo:
    command: ["echo", "foo"]
```    

```yaml
services:
  foo:
    command: ["echo", "bar"]
```

> MUST result in a Compose application model equivalent to yaml tree:

Reviewer 2: ```suggestion
MUST result in a Compose application model equivalent to the YAML tree:
```

```yaml
services:
  foo:
    DNS:
      command: ["echo", "bar"]
```

> ### Resources with unicity

Reviewer 3: ```suggestion
### Unique resources
```

Service's [ports](#ports), [volumes](#volumes), [secrets](#secrets) and [configs](#configs). While those

> are modeled in a Compose file as sequence, they MUST be unique regarding target resource in container:

Reviewer 2: ```suggestion
are modeled in a Compose file as a sequence, they MUST be unique regarding the target resource in container:
```

| attribute   | unique key               |
|-------------|--------------------------|
| volumes     |  target                  |
| secrets     |  source                  |
| configs     |  source                  |
| ports       |  {ip, target, published, protocol}   |

> While merging Compose files, a Compose Implementation MUST respect this unique keys and merge entries

Reviewer 2: ```suggestion
While merging Compose files, a Compose implementation MUST respect these unique keys and merge the entries
```

in sequence just like this would have been a mapping.

> Merging those two example yaml trees:

Reviewer 2: ```suggestion
Merging those two example YAML trees:
```

```yaml
services:
  foo:
    volumes:
      - foo:/work
```    

```yaml
services:
  foo:
    volumes:
      - bar:/work
```

> MUST result in a Compose application model equivalent to yaml tree:

Reviewer 2: ```suggestion
MUST result in a Compose application model equivalent to the YAML tree:
```
